sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/StarJoinCostBasedReorderSuite.scala

```diff
@@ -351,6 +351,18 @@ class StarJoinCostBasedReorderSuite extends PlanTest with StatsEstimationTestBas
         .join(t5.join(t6, Inner, Some(nameToAttr("t5_c2") === nameToAttr("t6_c2"))), Inner,
           Some(nameToAttr("d2_c2") === nameToAttr("t5_c1")))
         .select(outputsOf(d1, t3, t4, f1, d2, t5, t6, d3, t1, t2): _*)
+    } else {
```

Comment:
Why does this differ? we'd rather not allow different user-visible behavior

Reply:
I think the difference is caused by the 'toMap' method, the code affected for this case includes https://github.com/apache/spark/blob/c280c7f529e2766dd7dd45270bde340c28b9d74b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L153 to https://github.com/apache/spark/blob/c280c7f529e2766dd7dd45270bde340c28b9d74b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L155 and the result of `searchLevel` method (https://github.com/apache/spark/blob/c280c7f529e2766dd7dd45270bde340c28b9d74b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L208 ~ https://github.com/apache/spark/blob/c280c7f529e2766dd7dd45270bde340c28b9d74b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L234)

Reply:
For this case:

In Scala 2.12, the plans level 0 ~ level 4 of `foundPlans` are
```
level 0: (t4 ), (d2 ), (t5 ), (d3 ), (d1 ), (f1 ), (t2 ), (t6 ), (t1 ), (t3 )
level 1: {t5 t6 }, {t4 t3 }, {d3 f1 }, {t2 t1 }, {d2 f1 }, {d1 f1 }
level 2: {d2 d1 f1 }, {d2 d3 f1 }, {d3 d1 f1 }
level 3: {d2 d1 d3 f1 }
```

In the level 3, the `newJoinPlan` produce order are:

```
Join Inner, (d1_pk#5 = f1_fk1#0)
:- Join Inner, (f1_fk2#1 = d2_pk#8)
:  :- Join Inner, (f1_fk3#2 = d3_pk#11)
```

then 

```
Join Inner, (f1_fk2#1 = d2_pk#8)
:- Join Inner, (d1_pk#5 = f1_fk1#0)
:  :- Join Inner, (f1_fk3#2 = d3_pk#11)
```

The Cost of 2 candidates both `Cost(200,9200)` and we choose the first generated one, so the expected result is 
```
 f1.join(d3, Inner, Some(nameToAttr("f1_fk3") === nameToAttr("d3_pk")))
        .join(d1, Inner, Some(nameToAttr("f1_fk1") === nameToAttr("d1_pk")))
        .join(d2, Inner, Some(nameToAttr("f1_fk2") === nameToAttr("d2_pk")))
...
```

Reply:
In Scala 2.13, the plans level 0 ~ level 4 of `foundPlans` are
```
level 0: (t5) (d3) (d1) (f1) (t2) (d2) (t6) (t3) (t4) (t1)
level 1:  {d3 f1} {t2 t1} {d2 f1} {t5 t6} {d1 f1} {t4 t3}
level 2: {d2 d1 f1} {d3 d1 f1} {d2 d3 f1}
level 3: {d1 d2 d3 f1} 
```

In the level 3, the newJoinPlan produce order are:

```
Join Inner, (f1_fk2#1 = d2_pk#8)
:- Join Inner, (d1_pk#5 = f1_fk1#0)
:  :- Join Inner, (f1_fk3#2 = d3_pk#11)
```
then

```
Join Inner, (d1_pk#5 = f1_fk1#0)
:- Join Inner, (f1_fk2#1 = d2_pk#8)
:  :- Join Inner, (f1_fk3#2 = d3_pk#11)
```

The Cost of 2 candidates both Cost(200,9200) and we choose the first generated one too , so the expected result change to 

```
 f1.join(d3, Inner, Some(nameToAttr("f1_fk3") === nameToAttr("d3_pk")))
        .join(d2, Inner, Some(nameToAttr("f1_fk2") === nameToAttr("d2_pk")))
        .join(d1, Inner, Some(nameToAttr("f1_fk1") === nameToAttr("d1_pk")))
```

Reply:
Maybe we can add a Manual sort like 
```
    while (k <= lev - k) {
      val oneSideCandidates = existingLevels(k).values.toSeq.sortBy(_.itemIds.head)
      for (i <- oneSideCandidates.indices) {
        val oneSidePlan = oneSideCandidates(i)
        val otherSideCandidates = if (k == lev - k) {
          // Both sides of a join are at the same level, no need to repeat for previous ones.
          oneSideCandidates.drop(i)
        } else {
          existingLevels(lev - k).values.toSeq.sortBy(_.itemIds.head)
        }
..
```
to ensure the candidates plan input order, but the results tend to be 2.13

Reply:
Yeah is there any way to define an ordering on the outputs we are asserting about? assuming their order doesn't matter, then it's OK to sort what comes back into the test. That would be a tidier resolution.

Reply:
Address [b8608d8](https://github.com/apache/spark/pull/29434/commits/b8608d8b4b963a5039bf2cef680a4243f74144f7) add a sort to ensure the traversal order of candidate plans.